extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSource.java

```diff
@@ -211,10 +215,31 @@ protected InputEntity createEntity(CloudObjectLocation location)
   @Override
   protected Stream<InputSplit<List<CloudObjectLocation>>> getPrefixesSplitStream(@Nonnull SplitHintSpec splitHintSpec)
   {
-    final Iterator<List<S3ObjectSummary>> splitIterator = splitHintSpec.split(
-        getIterableObjectsFromPrefixes().iterator(),
-        object -> new InputFileAttribute(object.getSize())
-    );
+    final Iterator<List<S3ObjectSummary>> splitIterator;
+
+    // Skip files that didn't match filter from splitHintSpec.
+    if (org.apache.commons.lang.StringUtils.isNotBlank(getFilter())) {
```

Comment:
Sorry to play spoilsport but I see one potential issue with this PR. :(
IMHO we should push the filtering to getIterableObjectsFromPrefixes(). This way the filtered path will not be used in splitHintSpecCalculations. 
This logic may not work with people who have customer splitSpecs which do not take size into account. 

Also, I think we add the filtering to `getIterableObjectsFromPrefixes` in other input sources also like
1. AzureInputSource
2. OSSInputSource
3. GCSInputSource

we will not have the subtle bug where the path is used for calculating splits but is ultimately filtered out when returning to the user. 

@didip Appreciate your patience in answering the questions.

Reply:
@cryptoe ok, I believe I have addressed your concern. In the latest codebase I moved the filtering logic into the respective `getIterableObjectsFromPrefixes()`. 

This cleaned up the code a lot, which I think address the readability problem that @kfaraz raised.

All of my new test cases still passed.

Reply:
Good catch @cryptoe